add mixscape function

xinyuejohn opened the pull request:
<!-- Many thanks for contributing to this project! -->

**PR Checklist**

<!-- Please fill in the appropriate checklist below (delete whatever is not relevant). These are the most common things requested on pull requests (PRs). -->

-   [ ] Referenced issue is linked
-   [ ] If you've fixed a bug or added code that should be tested, add tests!
-   [ ] Documentation in `docs` is updated

**Description of changes**

<!-- Please state what you've changed and how it might affect the user. -->

**Technical details**

<!-- Please state any technical details such as limitations, reasons for additional dependencies, benchmarks etc. here. -->

**Additional context**

<!-- Add any other context or screenshots here. -->

Zethson commented on pertpy/tools/_mixscape.py:
```diff
@@ -0,0 +1,225 @@
+    return [mu, sd]
+
+
+def mixscape(
```
The function has a few unused arguments so far. We should ensure that the final implementation uses all of them.

Zethson commented on pertpy/tools/_mixscape.py:
```diff
@@ -0,0 +1,225 @@
+    fine_mode: bool | None = False,
+    fine_mode_labels: str | None = "guide_ID",
+    prtb_type: str | None = "KO",
+    copy: bool | None = False,
```
Maybe a better idea to only allow bool -> True/False and set the default to False? I don't see why a None value would make sense here.

Zethson commented on pertpy/tools/_mixscape.py:
```diff
@@ -0,0 +1,225 @@
+                de_genes = np.array([])
+            perturbation_markers[(category, gene)] = de_genes
+
+    all_markers = [list(perturbation_markers[key]) for key in perturbation_markers.keys()][0]
```
Could this block of code get a comment above it? What and why are we doing this? Just 1 sentence.

Zethson commented on pertpy/tools/_mixscape.py:
```diff
@@ -0,0 +1,225 @@
+    all_markers = [list(perturbation_markers[key]) for key in perturbation_markers.keys()][0]
+    missing_genes = []
+    for marker in all_markers:
+        if marker not in set(adata.var["name"]):
```
adata.var_names could also work?

xinyuejohn replied:
yes, it also works

Zethson commented on pertpy/tools/_mixscape.py:
```diff
@@ -0,0 +1,225 @@
+        if marker not in set(adata.var["name"]):
+            missing_genes.append(marker)
+
+    adata_comp = adata
```
Could this block of code get a comment above it? What and why are we doing this? Just 1 sentence.

Zethson commented on pertpy/tools/_mixscape.py:
```diff
@@ -0,0 +1,225 @@
+                de_genes_indices = _get_column_indices(adata, list(de_genes))
+                dat = X[all_cells][:, de_genes_indices]
+
+                """
```
What about this comment from the R code?

Zethson commented on pertpy/tools/_mixscape.py:
```diff
@@ -0,0 +1,225 @@
+                    lik_ratio = proba[:, 0] / proba[:, 1]
+                    post_prob = 1 / (1 + lik_ratio)
+
+                    adata.obs[new_class_name][[orig_guide_cells_index[a] for a in np.where(post_prob > 0.5)[0]]] = gene
```
Can we replace `a` with something more descriptive?

Zethson commented on pertpy/tools/_mixscape.py:
```diff
@@ -0,0 +1,225 @@
+
+                    adata.obs[new_class_name][[orig_guide_cells_index[a] for a in np.where(post_prob > 0.5)[0]]] = gene
+                    adata.obs[new_class_name][
+                        [orig_guide_cells_index[a] for a in np.where(post_prob <= 0.5)[0]]
```
What is `a`?

Zethson reviewed (commented):
Very strong work, impressive!

Have some minor comments.

Zethson commented on pertpy/plot/_mixscape_barplot.py:
```diff
@@ -0,0 +1,90 @@
+        save: If True or a str, save the figure. A string is appended to the default filename. Infer the filetype if ending on {'.pdf', '.png', '.svg'}.
+
+    Returns:
+        If show is False, return ggplot object used for drawn.
```
> used for drawn
This reads a bit wrong. `used to draw the plot` or something might be better.

Zethson commented on pertpy/plot/_mixscape_barplot.py:
```diff
@@ -0,0 +1,90 @@
+    count = pd.crosstab(index=adata.obs[mixscape_class_global], columns=adata.obs[control])
+    df = count / count.sum()
+    df2 = pd.melt(df, ignore_index=False).reset_index()
+    test = df2[df2[mixscape_class_global] == "KO"]
```
test? Not the best variable name....

Zethson commented on pertpy/plot/_mixscape_barplot.py:
```diff
@@ -0,0 +1,90 @@
+    test = test.sort_values("value", ascending=False)
+
+    new_levels = test[control]
+    df2[control] = pd.Categorical(df2[control], categories=new_levels, ordered=False)
```
It's not perfectly clear here what "df", "df2" and "df3" represent. Could you please name them better?

Zethson commented on pertpy/plot/_mixscape_heatmap.py:
```diff
@@ -0,0 +1,49 @@
+    save: bool | str | None = None,
+    **kwds,
+):
+    """Heatmap plot using mixscape results. Need to run `pt.tl.mixscape` first.
```
```suggestion
    """Heatmap plot using mixscape results. Requires `pt.tl.mixscape()` to be run first.
```

Zethson commented on pertpy/plot/_mixscape_heatmap.py:
```diff
@@ -0,0 +1,49 @@
+
+    """
+    adata_subset = adata[(adata.obs[labels] == target_gene) | (adata.obs[labels] == control)].copy()
+    sc.tl.rank_genes_groups(adata_subset, layer=layer, groupby=labels, method="wilcoxon")
```
The method should be parameterized with wilcoxon being the default parameter.

Zethson commented on pertpy/plot/_mixscape_heatmap.py:
```diff
@@ -0,0 +1,49 @@
+    adata_subset = adata[(adata.obs[labels] == target_gene) | (adata.obs[labels] == control)].copy()
+    sc.tl.rank_genes_groups(adata_subset, layer=layer, groupby=labels, method="wilcoxon")
+    sc.pp.scale(adata_subset, max_value=vmax)
+    sc.pp.subsample(adata_subset, n_obs=900)
```
The subsample number here is very arbitrary and a magic number. This should be paramterized. Why is it 900? I assume that this produced the closest plot to the original pipeline? Then we should use the number 900 in our tutorial, but still.

Zethson commented on pertpy/plot/_mixscape_violin.py:
```diff
@@ -0,0 +1,204 @@
+    ax: Axes | None = None,
+    **kwds,
+):
+    """Violin plot using mixscape results. Need to run `pt.tl.mixscape` first.
```
```suggestion
    """Violin plot using mixscape results. Requires `pt.tl.mixscape` to be run first.
```

Zethson commented on pertpy/plot/_mixscape_violin.py:
```diff
@@ -0,0 +1,204 @@
+            mixscape_class_mask |= adata.obs[groupby] == ident
+    adata = adata[mixscape_class_mask]
+
+    import seaborn as sns  # Slow import, only import if called
```
I am surprised that the seaborn import is slow? Think it's mostly just wrappers of a few matplotlib functions?

xinyuejohn replied:
No idea, I copied this from scanpy.pl.violin directly.

Zethson commented on pertpy/plot/_plotperturbscore.py:
```diff
@@ -0,0 +1,163 @@
+from __future__ import annotations
```
The naming of the python file and the function is not that great. See above.

Zethson commented on pertpy/plot/_plotperturbscore.py:
```diff
@@ -0,0 +1,163 @@
+)
+
+
+def plotperturbscore(
```
Naming

Zethson commented on pertpy/plot/_plotperturbscore.py:
```diff
@@ -0,0 +1,163 @@
+    before_mixscape=False,
+    perturbation_type: str = "KO",
+):
+    """Density plots to visualize perturbation scores calculated from RunMixscape function.
```
```suggestion
    """Density plots to visualize perturbation scores calculated by the `pt.tl.mixscape` function.
```

Zethson commented on pertpy/plot/_plotperturbscore.py:
```diff
@@ -0,0 +1,163 @@
+        split_by: Provide the column `.obs` if multiple biological replicates exist to calculate
+            the perturbation signature for every replicate separately.
+        before_mixscape: Option to split densities based on mixscape classification (default) or original target gene classification. Default is set to NULL and plots cells by original class ID.
+        perturbation_type: specify type of CRISPR perturbation expected for labeling mixscape classifications. Default is KO.
```
Missing empty line before returns

Zethson commented on pertpy/plot/_plotperturbscore.py:
```diff
@@ -0,0 +1,163 @@
+    Returns:
+        The ggplot object used for drawn.
+    """
+    perturbation_score = None
```
Big blob of code afterwards. Maybe 1-2 sections + 1 line comments can help digest this?

xinyuejohn replied:
No problem

Zethson reviewed (changes requested):
Thank you very much!

1. This is a bigger request, but I'd love to see all mixscape related functions including the perturbation space (that @Koncopd implemented) move into a Mixscape class like Augurpy. Also all plotting functions. This will of course also change all imports etc. But to stay consistent with the other tools this might be worth the effort.
2. In general there's no need for an empty line after the function docstring. Please remove those :)
3. The mixscape plotting functions have the requirement for the mixscape function to be run. Makes sense of course. Can we add a small function for every plot call that checks for that beforehand? The mixscape function adds some stuff to the AnnData object and we can just check for the presence of those. If not present we can throw a custom Error.
4. Could I get a quick runtime comparison of the R vs our Python implementation, please?

xinyuejohn commented on docs/usage/usage.md:
```diff
@@ -118,3 +119,15 @@ of each cell.
+    :toctree: plot
+
+    plot.mixscape_violin
+    plot.plotperturbscore
```
What about "mixscape_perturbscore"

Zethson replied:
Sure.

xinyuejohn commented:
> Thank you very much!
> 
> 1. This is a bigger request, but I'd love to see all mixscape related functions including the perturbation space (that @Koncopd implemented) move into a Mixscape class like Augurpy. Also all plotting functions. This will of course also change all imports etc. But to stay consistent with the other tools this might be worth the effort.
> 2. In general there's no need for an empty line after the function docstring. Please remove those :)
> 3. The mixscape plotting functions have the requirement for the mixscape function to be run. Makes sense of course. Can we add a small function for every plot call that checks for that beforehand? The mixscape function adds some stuff to the AnnData object and we can just check for the presence of those. If not present we can throw a custom Error.
> 4. Could I get a quick runtime comparison of the R vs our Python implementation, please?

Thanks for your comments! I changes the code accordingly:
1. A Mixscape class has been added and the original mixscape function still use the original name.  And mixscape function has been split up and the max-complexity in flake8 has been changed to 15 again.
2. Removed
3. Added for plot functions except violin plot. Because mixscape_violin can also be used for violin plotting with part of mixscape results. For example, using mixscape_violin for PD-L1 protein expression used mdata['adt'] and it used partial mixscape results copied from mdata['rna'].
4. FYI, mixscape in R is ~137.5 s, in pertpy is ~15.4 s.
